fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/lock/InMemoryResourceLockManager.java

```diff
@@ -29,67 +35,86 @@ public class InMemoryResourceLockManager implements ResourceLockManager {
+                    // Can't get an exclusive lock on a resource that already has any lock not owned by the current
+                    // transaction or get any lock if there is already an exclusive lock. If the current transaction
+                    // held this exclusive lock we would not have gotten this far.
+                       throw new ConcurrentUpdateException(
+                               String.format("Cannot update %s because it is being updated by another transaction.",
+                                       resourceIdStr));
+                }
             }
 
             LOG.debug("Transaction {} acquiring lock on {}", txId, resourceIdStr);
 
-            lockedResources.add(resourceIdStr);
-            transactionLocks.computeIfAbsent(txId, key -> Sets.newConcurrentHashSet())
-                    .add(resourceIdStr);
+            final var resourceLock = new ResourceLockImpl(lockType, txId, resourceIdStr);
+            resourceLocks.computeIfAbsent(resourceIdStr, key -> Sets.newConcurrentHashSet()).add(resourceLock);
+            transactionLocks.computeIfAbsent(txId, key -> Sets.newConcurrentHashSet()).add(resourceLock);
         }
     }
 
     @Override
     public void releaseAll(final String txId) {
         final var locks = transactionLocks.remove(txId);
         if (locks != null) {
-            locks.forEach(resourceId -> {
-                LOG.debug("Transaction {} releasing lock on {}", txId, resourceId);
-                synchronized (acquireInternalLock(resourceId)) {
-                    lockedResources.remove(resourceId);
-                    internalResourceLocks.remove(resourceId);
+            locks.forEach(lock -> {
+                LOG.debug("Transaction {} releasing lock on {}", txId, lock);
+                synchronized (acquireInternalLock(txId, lock.getResourceId(), lock.getLockType())) {
+                    internalResourceLocks.get(lock.getResourceId()).removeIf(
+                            l -> l.getTransactionId().equals(txId));
+                    if (internalResourceLocks.get(lock.getResourceId()).size() == 0) {
+                        internalResourceLocks.remove(lock.getResourceId());
+                    }
+                    resourceLocks.get(lock.getResourceId()).removeIf(
+                            l -> l.getTransactionId().equals(txId));
+                    if (resourceLocks.get(lock.getResourceId()).size() == 0) {
+                        resourceLocks.remove(lock.getResourceId());
+                    }
                 }
             });
         }
     }
 
-    private Object acquireInternalLock(final String resourceId) {
-        return internalResourceLocks.computeIfAbsent(resourceId, key -> new Object());
+    private Object acquireInternalLock(final String txId, final String resourceId, final ResourceLockType lockType) {
+        return internalResourceLocks.computeIfAbsent(resourceId,
```

Comment:
I don't fully understand what you mean. If a thread calls this method to acquire the lock defined by the key resourceID how does storing a `ResourceLock` differ from storing a `new Object()`. I don't see how the old code actually stopped a second thread from acquiring the same lock.

Reply:
The original code was synchronizing on a single lock instance per resource. You can synchronize on a ResourceLock instance instead, if you want, but the important part is that the synchronization happens on the same exact instance in all threads. Otherwise, the lock won't work.

That aside, I don't have this code pulled down currently, but it seems like this method might be returning a boolean now, which is the output of `Set.add()`. If that is true, it actually means that the current code is synchronizing across all threads regardless of the resource they're attempting to lock.

Reply:
If I am synchronizing on a single ResourceLock instance for each resource, won't that mean I can't have concurrent non-exclusive locks? Do I need to have the ResourceLock contain all the locks for a resource? i.e.
```java
class ResourceLock {
    private String resourceId;
    private Map<String, ResourceLockType> transactionLocks;
    ...
```

Reply:
No, for the internal lock it doesn't actually matter what the object is, which is why it was originally using `Object`. The point of the internal lock is so that one thread can safely acquire the resource lock without a race condition. The internal lock is only held long enough for a thread to either acquire the resource lock that it's attempting to acquire or to find out that it's not allowed to have a lock on the resource.

For example, take the following alternate implementation:

```java
public class InMemoryResourceLockManager {
    private synchronized void acquireInternal(final String txId, final FedoraId resourceId, final ResourceLockType lockType) {
     // code goes here
    }
}
``` 

In that example, Java will only allow a single thread to enter the `acquireInternal()` method at a time. That code is equivalent to:

```java
public class InMemoryResourceLockManager {
    private void acquireInternal(final String txId, final FedoraId resourceId, final ResourceLockType lockType) {
        syncronized (this) {
            // code goes here
         }
    }
}
```

You could write the lock manager like that, and it would work correctly. The downside of that approach is that every thread would block on that method regardless of the resource they were trying to lock. So, instead of using a single internal lock guard lock acquisition for all resource, the intent was to use a per-resource internal lock, so that there is only contention between threads that are attempting to lock the same resource. To continue the example, this is the equivalent of:

```java
public class InMemoryResourceLockManager {
    private void acquireInternal(final String txId, final FedoraId resourceId, final ResourceLockType lockType) {
        syncronized (internalResourceLocks.computeIfAbsent(resourceId, key -> new Object())) {
            // code goes here
         }
    }
}
```

The point is, that it is only safe for one thread at a time to be executing the logic that awards or declines a lock on a resource, so you need a way to synchronize the lock acquisition code.

Does that make sense?

Reply:
That said, I think there's a bug in the way the original code worked that could result in two different threads independently holding different internal locks on the same resource...

I'll think about it a little more this weekend. For sure the safest thing to do would be to globally synchronize lock acquisition and release. It would perform a little worse than synchronizing by resource, but it certainly wouldn't be the largest latency source in Fedora.